File: assignment_2/rrt.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <math.h>
#include <random>
#include "graph.h"
using namespace std;

// Uniform random number generator in [-50,50]
random_device rd;
mt19937 gen(rd());
uniform_real_distribution<> uniform_dis(-50.0, 50.0);

// Start and goal nodes given in each problem
double eps [5] = {10,5,5,2,1};  
double x_start [5] = {0,27,45,-16,39};
double y_start [5] = {0,30,-45,10,5};
double goal_region [5][3] = {{-38,20,10},{-48,20,10},{-45,45,15},{18,-45,5},{38,-8,3}};

// See if point has reached the goal
bool reachedGoal(double x, double y, double x_goal, double y_goal, double r_goal)
{
    if(pow((x - x_goal),2) + pow((y - y_goal),2) < pow(r_goal,2))
    {
        return true;
    }
    else 
    {
        return false;
    }
}

// Check that no part of line segment intersects an obstacle
bool isFree(double v_x, double v_y, double w_x, double w_y, const char* obsFilename)
{
    fstream obsFile(obsFilename);
    string obsLine;

    // Get number of obstacles
    getline(obsFile,obsLine);
    int numObs = stoi(obsLine);

    // Create obstacle array
    double obs[numObs][3];

    // Store values line by line (each line is one obstacle)
    for(int i=0; i<numObs; i++)
    {
        getline(obsFile, obsLine);
        stringstream obsVal(obsLine);
        string value;
        for(int j=0; j<3; j++)
        {
            getline(obsVal, value, ',');
            obs[i][j] = stod(value);
        }
    }
    obsFile.close();

    // Check if the point (x,y) is in each obstacle
    for(int i=0; i<numObs; i++)
    {
        // Find minimum distance of obstacle center to line segment using dot product
        // Reference: https://www.geeksforgeeks.org/minimum-distance-from-a-point-to-the-line-segment-using-vectors/
        double vw_x = w_x - v_x;
        double vw_y = w_y - v_y;
        double wc_x = obs[i][0] - w_x;
        double wc_y = obs[i][1] - w_y;
        double vc_x = obs[i][0] - v_x;
        double vc_y = obs[i][1] - v_y;

        double vw_dot_wc = vw_x*wc_x + vw_y*wc_y;
        double vw_dot_vc = vw_x*vc_x + vw_y*vc_y;

        double minDistance;

        if(vw_dot_wc > 0)
        {
            minDistance = sqrt(pow(wc_x,2) + pow(wc_y,2));
        }
        else if(vw_dot_vc < 0)
        {
            minDistance = sqrt(pow(vc_x,2) + pow(vc_y,2));
        }
        else
        {
            double vwLength = sqrt(pow(vw_x,2) + pow(vw_y,2));
            minDistance = abs(vw_x*vc_y - vw_y*vc_x)/vwLength;
        }

        // Check if minimum distance to segment is smaller than the obstacle radius
        if(minDistance<obs[i][2])
        {
            return false;
        }
    }

    return true;
}

// Obstacle detection function (only checks endpoints)
bool isFreeLite(double x, double y, const char* obsFilename)
{
    fstream obsFile(obsFilename);
    string obsLine;

    // Get number of obstacles
    getline(obsFile,obsLine);
    int numObs = stoi(obsLine);

    // Create obstacle array
    double obs[numObs][3];

    // Store values line by line (each line is one obstacle)
    for(int i=0; i<numObs; i++)
    {
        getline(obsFile, obsLine);
        stringstream obsVal(obsLine);
        string value;
        for(int j=0; j<3; j++)
        {
            getline(obsVal, value, ',');
            obs[i][j] = stod(value);
        }
    }
    obsFile.close();

    // Check if the point (x,y) is in each obstacle
    for(int i=0; i<numObs; i++)
    {
        if(pow((x - obs[i][0]),2) + pow((y - obs[i][1]),2) < pow(obs[i][2],2))
        {
            // The point intersects an obstacle
            return false;
        }
    }

    // The point does not intersect any of the obstacles
    return true;
}

int main()
{
    for(int j = 0; j < 5; j++)
    {
        // Maximum number of attempts before timing out
        int maxNodes = 5000;

        // Allocate space for u,v,w
        double u_x, u_y, v_x, v_y, w_x, w_y;

        // Initialize graph with start node
        Graph G(x_start[j],y_start[j]);

        // Iterate until maximum nuber of attempts reached
        for(int i = 0; i<maxNodes; i++)
        {
            // Choose point from uniform random distribution
            u_x = uniform_dis(gen); double u_y = uniform_dis(gen);

            // Get node closest to random point
            Node* nearNode = G.nearestNode(u_x,u_y);
            v_x = nearNode->x; v_y = nearNode->y;

            // Episilon step toward the random point (if epsilon is sufficiently small)
            double length = sqrt(pow((u_x - v_x), 2) + pow((u_y - v_y), 2));
            double ratio;
            if (eps[j]<length)
            {
                ratio = eps[j]/length;
            }
            else
            {
                ratio = 1.0;
            }
            w_x = ratio*(u_x-v_x) + v_x;
            w_y = ratio*(u_y-v_y) + v_y;

            // // works approximately
            // if(isFreeLite(w_x,w_y,"obstacles.txt"))
            // {
            //     G.newNode(w_x,w_y,nearNode);
            // }

            // Check that path segment does not intersect obstacles
            if(isFree(v_x,v_y,w_x,w_y,"obstacles.txt"))
            {
                // Add new point to graph
                G.newNode(w_x,w_y,nearNode);

                // End while loop if goal region is reached
                if(reachedGoal(w_x,w_y,goal_region[j][0],goal_region[j][1],goal_region[j][2]))
                {
                    cout << i << "\n";
                    cout << "Goal reached! \n";
                    break;
                }
            }
        }

        // Warning message if goal region is not reached in alotted time
        if(!reachedGoal(w_x,w_y,goal_region[j][0],goal_region[j][1],goal_region[j][2]))
        {
            cout << "Timed out: goal not reached! \n";
        }

        // Save path and search tree to files
        string outputPath = "outputs/output_path_" + to_string(j+1) + ".txt";
        string outputTree = "outputs/output_tree_" + to_string(j+1) + ".txt";
        const char *OutputPath = outputPath.c_str();
        const char *OutputTree = outputTree.c_str();

        // Clear output files if they already exist
        remove(OutputPath);
        remove(OutputTree);

        // Write path and search tree to text files
        G.savePath(OutputPath,G.nodes.back());
        G.saveSearchTree(OutputTree);
    }

    return 0;
}

File: assignment_3/RRT.py
import numpy as np
from numpy import genfromtxt
import os

# Given problem parameters
delta_t = 0.1       # ensures collision detection happens before delta > 0.5
eps = [10,5,5,3,2]  
x_start = [30,40,30,-50,-30]
y_start = [-35,-40,40,-30,-35]
theta_start = [np.pi/2,0,3*np.pi/2,np.pi/2,np.pi/4]
goal_region = [[0,0,5],[30,40,5],[-50,-30,5],[30,-35,5],[-35,30,5]]

# Parse obstacle file given
obsFile = 'obstacles.txt'
obs = np.asarray(genfromtxt(obsFile, delimiter=',',skip_header=True))
numObs = obs.shape[0]

# Parse vehicle file given
vehicleFile = 'H3_robot.txt'
vehicle = np.asarray(genfromtxt(vehicleFile, delimiter=','))
numVehiclePts = vehicle.shape[0]
vehicleRadius = 1 # max radial distance from center for quick collision detection

# Waypoints searched by RRT
class Node:
    def __init__(self,x,y,theta,v,omega,t,nearestNode):

        # State variables
        self.x = x
        self.y = y
        self.theta = theta
        self.v = v
        self.omega = omega
        self.t = t

        # Data structure maintainace
        self.parentNode = nearestNode
        self.numOutgoingEdges = 0
        self.outgoingEdges = []
        self.inPath = False

# Dynamics and trajectories between waypoints
class Edge:
    def __init__(self,startNode,endNode,a,alpha,xTraj,yTraj,thetaTraj,vTraj,omegaTraj,tTraj):

        # Pointers to startpoint and endpoint of edge
        self.startNode = startNode
        self.endNode = endNode
        self.inPath = False

        # Control inputs - constant over edge
        self.a = a
        self.alpha = alpha
        
        # Store state values
        self.xTraj = xTraj
        self.yTraj = yTraj
        self.thetaTraj = thetaTraj
        self.vTraj = vTraj
        self.omegaTraj = omegaTraj
        self.tTraj = tTraj

# Store waypoints and dynamics in graph structure
class Graph:
    def __init__(self,x_start,y_start,theta_start):
        self.nodes = [Node(x_start,y_start,theta_start,0,0,0,None)]
        self.edges = []
        self.numNodes = 1
        self.numEdges = 0
    
    # Find existing node with smallest Euclidean distance to (x,y)
    def findNearestNode(self,x,y):
        distances = []
        for node in self.nodes:
            dist = np.sqrt((x-node.x)**2 + (y-node.y)**2)
            distances.append(dist)
        minIndex = np.argmin(distances)
        return self.nodes[minIndex]

    # Store a new node and edge
    def newNode(self,x,y,theta,v,omega,t,a,alpha,xTraj,yTraj,thetaTraj,vTraj,omegaTraj,tTraj,closeNode):
        newestNode = Node(x,y,theta,v,omega,t,closeNode)
        self.nodes.append(newestNode)
        self.edges.append(Edge(closeNode,newestNode,a,alpha,xTraj,yTraj,thetaTraj,vTraj,omegaTraj,tTraj))

        closeNode.numOutgoingEdges = closeNode.numOutgoingEdges + 1
        closeNode.outgoingEdges.append(newestNode)

        self.numNodes = self.numNodes + 1
        self.numEdges = self.numEdges + 1
    
    # Change attribute of nodes and edges in the feasible path found
    def definePath(self,goalNode):
        thisNode = goalNode
        while(thisNode != None):
            thisNode.inPath = True
            thisNode = thisNode.parentNode
        for edge in self.edges:
            if (edge.startNode.inPath & edge.endNode.inPath):
                edge.inPath = True
    
    # Save the feasible path to a .csv file
    def savePath(self,pathFile,x_start,y_start):
        x = [x_start]
        y = [y_start]
        for edge in self.edges:
            if(edge.inPath):
                x = np.append(x,edge.xTraj)
                y = np.append(y,edge.yTraj)
        path = np.vstack((x,y))
        np.savetxt(pathFile,path,delimiter=',')
    
    # Save the paths searched to a .csv file
    def saveSearchTraj(self,trajFile):
        x = [np.NaN]
        y = [np.NaN]
        for edge in self.edges:
            x = np.append(x,edge.xTraj)
            x = np.append(x,np.NaN)
            y = np.append(y,edge.yTraj)
            y = np.append(y,np.NaN)        
        searchTraj = np.vstack((x,y))
        np.savetxt(trajFile,searchTraj,delimiter=',')
    
    # Save the search tree in the format specified by the problem set
    def saveSearchTree(self,treeFile):
        search = np.zeros([self.numEdges,8])
        i = 0
        for edge in self.edges:
            search[i][0] = edge.startNode.t
            search[i][1] = edge.startNode.x
            search[i][2] = edge.startNode.y
            search[i][3] = edge.startNode.theta
            search[i][4] = edge.startNode.v
            search[i][5] = edge.startNode.omega
            search[i][6] = edge.a
            search[i][7] = edge.alpha
            i = i+1
        np.savetxt(treeFile,search,delimiter=',')

# Full collision checker - checks point by point
def isFree(x,y,theta):

    # Rotate and translate the vehicle
    c = np.cos(theta)
    s = np.sin(theta)
    rotMat = np.array(((c, s), (-s, c)))
    transMat = np.hstack((x*np.ones((numVehiclePts,1)),y*np.ones((numVehiclePts,1))))
    rotVehicle = np.dot(vehicle,rotMat)
    shiftVehicle = rotVehicle + transMat

    # Check for collision with each obstacle at each point
    for i in range(numVehiclePts):
        for j in range(numObs):
            if((shiftVehicle[i][0]-obs[j][0])**2 + (shiftVehicle[i][1]-obs[j][1])**2 < (obs[j][2])**2):
                return False
    return True

# Quick collision checker - collision reported if the obstacle is within maximum radius of the vehicle
def isFreeQuick(x,y,thetaQuick=0):

    # Check for collision with each obstacle
    for i in range(numObs):
        if((x-obs[i][0])**2 + (y-obs[i][1])**2 < (obs[i][2] + vehicleRadius)**2):
            return False
    return True

# Determine if end condition has been met
def reachedGoal(x,y,x_goal,y_goal,r_goal):
    if ((x-x_goal)**2 + (y-y_goal)**2 < r_goal**2):
        return True
    else:
        return False

# Starting from nearestNode, generate a feasible trajectory and store it in the graph
def newTrajectory(G,nearestNode,eps):

    # Preallocate space for trajectories
    x_traj = []
    y_traj = []
    theta_traj = []
    v_traj = []
    omega_traj = []
    t_traj = []

    # State at nearestNode
    x_i = nearestNode.x
    y_i = nearestNode.y
    theta_i = nearestNode.theta
    v_i = nearestNode.v
    omega_i = nearestNode.omega
    t_i = nearestNode.t
    
    # To reduce the chance of exceeding maximum velocities, accelarations are
    # chosen 'randomly' to reduce the magnitude of the velocities

    # Choose a 'random' feasible linear accelaration
    if(v_i > 2.5):
        a = np.random.uniform(-2.0,0,1)
    elif(v_i < -2.5):
        a = np.random.uniform(0,2.0,1)
    else:
        a = np.random.uniform(-2.0,2.0,1)

    # Choose a 'random' feasible rotational accelaration
    if(omega_i > np.pi/4):
        alpha = np.random.uniform(-np.pi/2,0,1)
    elif(omega_i < -np.pi/4):
        alpha = np.random.uniform(0,np.pi/2,1)
    else:
        alpha = np.random.uniform(-np.pi/2,np.pi/2,1)

    # Do not exceed a Euclidean distance of epsilon between waypoints
    while((x_i-nearestNode.x)**2 + (y_i-nearestNode.y)**2 < eps**2):

        # Apply Euler's method to approximate next state
        x_i1 = x_i + v_i*np.cos(theta_i)*delta_t
        y_i1 = y_i + v_i*np.sin(theta_i)*delta_t
        theta_i1 = theta_i + omega_i*delta_t
        v_i1 = v_i + a*delta_t
        omega_i1 = omega_i + alpha*delta_t
        t_i1 = t_i + delta_t
        
        # Break conditions
        if(abs(v_i1) > 5.0):
            break
        if(abs(omega_i1) > np.pi/2):
            break
        if(not isFree(x_i1,y_i1,theta_i1)):
            break

        # Update values
        x_i = float(x_i1)
        y_i = float(y_i1)
        theta_i = float(theta_i1)
        v_i = float(v_i1)
        omega_i = float(omega_i1)
        t_i = float(t_i1)

        # Store trajectory
        x_traj.append(x_i)
        y_traj.append(y_i)
        theta_traj.append(theta_i)
        v_traj.append(v_i)
        omega_traj.append(omega_i)
        t_traj.append(t_i)

    if(not (x_i-nearestNode.x)**2 + (y_i-nearestNode.y)**2 == 0): # only add node if vehicle can move
        G.newNode(x_i,y_i,theta_i,v_i,omega_i,t_i,a,alpha,x_traj,y_traj,theta_traj,v_traj,omega_traj,t_traj,nearestNode)
    else:
        pass

# RRT algorithm with vehicle dynamics and volume
def rrt():
    # Iterate over 5 problems given
    for j in range(5):

        # Initialize graph
        G = Graph(x_start[j],y_start[j],theta_start[j])

        # Limit runtime
        maxNodes = 20000
        for i in range(maxNodes):

            # Chose a random point in the plane
            u_x = np.random.uniform(-50.0,50.0,1)
            u_y = np.random.uniform(-50.0,50.0,1)

            # Choose the node closest to the random point
            nearNode = G.findNearestNode(u_x,u_y)

            # Start at the node, generate a random feasible trajectory, and store it in the graph
            newTrajectory(G,nearNode,eps[j])
            
            # End the search if the goal region has been reached
            if reachedGoal(G.nodes[-1].x,G.nodes[-1].y,goal_region[j][0],goal_region[j][1],goal_region[j][2]):
                print("Goal reached!")
                G.definePath(G.nodes[-1]) # Keep track of nodes and edges in feasbile path
                break
        
        # Warning statement if goal was not reached
        if(not reachedGoal(G.nodes[-1].x,G.nodes[-1].y,goal_region[j][0],goal_region[j][1],goal_region[j][2])):
            print("Timed out: goal not reached! \n")
        
        # File names for RRT search
        outputTree = "outputs/output_tree_" + str(j+1) + ".txt"
        outputPath = "outputs/output_path_" + str(j+1) + ".txt"
        outputTraj = "outputs/output_traj_"+ str(j+1) + ".txt"

        G.saveSearchTraj(outputTraj)                    # Save paths explored
        G.saveSearchTree(outputTree)                    # Save tree in form specified by problem set
        G.savePath(outputPath,x_start[j],y_start[j])    # Save feasible path found

# The important part
def main():
    rrt()

if __name__ == '__main__':
    main()